template/web/app/buy-me-coffee/_components/useSmartContractForms.ts

```diff
@@ -0,0 +1,115 @@
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { Abi, TransactionExecutionError } from 'viem';
+import { useSimulateContract, useWaitForTransactionReceipt, useWriteContract } from 'wagmi';
+import { UseContractReturn } from '@/hooks/contracts';
+import { useLoggedInUserCanAfford } from '@/hooks/useUserCanAfford';
+
+export enum TransactionStates {
+  START,
+  COMPLETE,
+  OUT_OF_GAS,
+}
+
+type AsyncFunction<Args extends unknown[], ReturnType> = (...args: Args) => Promise<ReturnType>;
+
+export default function useSmartContractForms({
+  gasFee,
+  contract,
+  name: functionName,
+  arguments: args,
+  enableSubmit: isValid,
+  reset,
+}: {
+  gasFee: bigint;
+  contract: UseContractReturn<Abi>;
+  name: string;
+  arguments: (number | string)[];
+  enableSubmit: boolean;
+  reset: AsyncFunction<unknown[], unknown>;
+}) {
+  const [transactionState, setTransactionState] = useState<TransactionStates | null>(null);
+
+  const canAfford = useLoggedInUserCanAfford(gasFee);
+
+  const { data: contractRequest } = useSimulateContract({
+    address: contract.status === 'ready' ? contract.address : undefined,
+    abi: contract.abi,
+    functionName: functionName,
+    args: args,
+    query: {
+      enabled: isValid && contract.status === 'ready',
+    },
+    value: gasFee,
+  });
+
+  const {
+    writeContract,
+    data: dataHash,
+    status: writeContractStatus,
+    error: writeContractError,
+  } = useWriteContract();
```

Comment:
Pretty sure we can just use `useWriteContract` instead of `useSimulateContract`. This PR is already big so I will tackle that next.

Reply:
Perhaps we need `useSimulateContract` to validate the form?

Reply:
`useSimulateContract` i believe was used for get the Gas fee.